sockeye/data_io.py

```diff
@@ -577,18 +572,17 @@ def _populate_bucket_batch_sizes(self):
                 batch_size_seq = self.batch_size
                 batch_size_word = batch_size_seq * average_seq_len
             self.bucket_batch_sizes[buck_idx] = (batch_size_seq, batch_size_word)
-        # Default bucket batch size: batch size of largest bucket batch by total elements, scaled for default bucket key
-        # Effectively, set default values to reserve enough space to provide a batch from any bucket
-        largest_total_batch_size = 0
+            # Track largest batch size by total elements
+            largest_total_batch_size = max(largest_total_batch_size, batch_size_seq * max(*bucket_shape))
+        # Final step: guarantee that largest bucket by sequence length also has largest total batch size.
```

Comment:
it might be worth pointing out that this is just a heuristic and does not guarantee full memory reuse across buckets (as this would require the default bucket to consist of both the maximum batch size and the maximum sentence length), which would allocate a lot of memory, but at the same time make sure all ndarrays from the default executor can be shared.

Reply:
Good idea to make that clear in the code!